numba/ir_utils.py

```diff
@@ -1854,3 +1854,68 @@ def raise_on_unsupported_feature(func_ir):
                "nopython-mode\n\nConflicting calls found at:\n %s")
         buf = '\n'.join([x.strformat() for x in gdb_calls])
         raise UnsupportedError(msg % buf)
+
+
+class InlineInlinables(object):
+
+    def __init__(self, func_ir, recursive=False):
+        self.func_ir = func_ir
+        self.recursive = recursive
+
+    def run(self):
+        """Run inlining of inlinables
+        """
+        if config.DEBUG:
+            print('before inline'.center(80, '-'))
+            print(self.func_ir.dump())
+            print(''.center(80, '-'))
+        modified = False
+        work_list = list(self.func_ir.blocks.items())
+        while work_list:
+            label, block = work_list.pop()
+            for i, instr in enumerate(block.body):
+                if isinstance(instr, ir.Assign):
+                    expr = instr.value
+                    if isinstance(expr, ir.Expr) and expr.op == 'call':
+                        if guard(self._do_work, work_list, block, i, expr):
+                            modified = True
+                            break # because block structure changed
+
+        if modified:
+            remove_dels(self.func_ir.blocks)
+            # repeat dead code elimination until nothing can be further
+            # removed
+            while (remove_dead(self.func_ir.blocks, self.func_ir.arg_names,
```

Comment:
Is dead code elimination necessary here? As discussed before, using it before typing is unsafe.